File: NCalculatorLib/Exprs/NumExpr.cs
using Null.Calculator;

#nullable enable

namespace NCalculatorLib.Exprs
{
    /// <summary>
    /// basic number
    /// 
    /// | num_literal
    /// </summary>
    public class NumExpr : Expr
    {
        public readonly string Value;

        public NumExpr(string value)
        {
            Value = value;
        }

        public override IEnumerable<Token> EnumTokens()
        {
            yield return new Token(TokenKind.Number, Value);
        }

        public override double Eval()
        {
            return double.Parse(Value);
        }
    }
}

File: NCalculatorCppLib/Exprs/num_expr.cpp
#include "NumExpr.h"

using namespace NCalculator;

namespace NCalculatorLibExprs
{

	NumExpr::NumExpr(const std::wstring &value) : Value(value)
	{
	}

	std::vector<token> NumExpr::enum_tokens()
	{
		std::vector<token> rst = std::vector<token>();
		rst.push_back(token(token_kind::Number, Value));
		return  rst;
	}

	double NumExpr::eval()
	{
		return std::stod(Value);
	}
}
